File: Tools/data/include_utils.cpp
//=============================================================================
//
// Adventure Game Studio (AGS)
//
// Copyright (C) 1999-2011 Chris Jones and 2011-2025 various contributors
// The full list of copyright holders can be found in the Copyright.txt
// file, which is part of this source code distribution.
//
// The AGS source code is provided under the Artistic License 2.0.
// A copy of this license can be found in the file License.txt and at
// https://opensource.org/license/artistic-2-0/
//
//=============================================================================
/// The "Include Utils" deal with the include/exclude list, containing patterns
/// in the .gitignore format, which may instruct to either include or
/// exclude particular file path.
/// The code is based on the Python's fnmatch, and a gist demonstrating
/// its port to C++, see:
/// https://github.com/python/cpython/blob/main/Lib/fnmatch.py
/// https://gist.github.com/alco/1869512
/// NOTE: the level of support for .gitignore format is not clear,
/// maybe not all of the possible edge cases are handled here.
/// For the reference, .gitignore specs:
/// https://git-scm.com/docs/gitignore
//=============================================================================
#include "data/include_utils.h"
#include "util/file.h"
#include "util/path.h"
#include "util/textstreamreader.h"
#include <regex>

using namespace AGS::Common;

namespace AGS
{
namespace DataUtil
{

using AGS::Common::Stream;

enum PatternType
{
    eInclude = 0,
    eExclude
};

struct Pattern
{
    PatternType Type;
    std::regex Regex;
    String TextualTranslatedPattern; // for debug purposes
    String TextualOriginalPattern; // for debug purposes
};

static bool is_regex_syntax_character(const int c)
{
    // see https://tc39.es/ecma262/#prod-SyntaxCharacter
    return (c == '^') || (c == '$') || (c =='\\') || (c == '.') || (c == '*') || (c == '+') || (c == '?') ||
        (c == '(') || (c == ')') || (c == '[') || (c == ']') || (c == '{') || (c == '}') || (c == '|');
}

static String translate_to_regex_string(const String &pattern)
{
    int i = 0;
    int n = (int)pattern.GetLength();
    String result;

    // The pattern must match the path section either at the beginning or right after the path separator
    if (!pattern.StartsWith("/"))
    {
        result.Append("(^|/)");
    }

    while (i < n) {
        char c = pattern[i];
        ++i;

        if (c == '*') {
            result.Append(".*");
        } else if (c == '?') {
            result.AppendChar('.');
        } else if (c == '[') {
            int j = i;
            /*
             * The following two statements check if the sequence we stumbled
             * upon is '[]' or '[!]' because those are not valid character
             * classes.
             */
            if (j < n && pattern[j] == '!')
                ++j;
            if (j < n && pattern[j] == ']')
                ++j;
            /*
             * Look for the closing ']' right off the bat. If one is not found,
             * escape the opening '[' and continue.  If it is found, process
             * the contents of '[...]'.
             */
            while (j < n && pattern[j] != ']')
                ++j;
            if (j >= n) {
                result.Append("\\[");
            } else {
                String stuff = pattern;
                stuff.ClipMid(i, j-i);
                stuff.Replace("\\"," \\\\");
                char first_char = pattern.GetAt(i);
                i = j + 1;
                result.Append("[");
                if (first_char == '!') {
                    result.Append("^");
                    stuff.ClipLeft(1);
                    result.Append(stuff);
                } else if (first_char == '^') {
                    result.Append("\\");
                    result.Append(stuff);
                } else {
                    result.Append(stuff);
                }
                result.Append("]");
            }
        } else {
            // Escape a regex syntax character if necessary, otherwise print one verbatim
            if (is_regex_syntax_character(c)) {
                result.Append("\\");
                result.AppendChar(c);
            } else {
                result.AppendChar(c);
            }
        }
    }

    // The pattern must match the path section either at the end or right before the path separator
    if (!pattern.EndsWith("/"))
    {
        result.Append("($|/)");
    }

    return result;
}

String normalize_separators_in_text(const String &path)
{
    String result = path;
    result.Replace("\\\\", "/");
    return result;
}

String normalize_separators_in_string(const String &path)
{
    String result = path;
    result.Replace("\\", "/");
    return result;
}

HError description_to_patterns(const std::vector<String> &description, std::vector<Pattern> &patterns)
{
    for (String line : description)
    {
        line.Trim();

        // line is empty or a comment
        if(line.IsEmpty() || line.StartsWith("#"))
            continue;

        Pattern p;
        p.Type = eInclude;
        if (line[0] == '!')
        {
            p.Type = eExclude;
            line.ClipLeft(1);
        }
        line.MakeLowerUTF8(); // for case insensitivity
        String shell_regex_txt = normalize_separators_in_text(line); // for unix paths

        String regex_txt = translate_to_regex_string(shell_regex_txt);

        p.TextualOriginalPattern = shell_regex_txt;
        p.TextualTranslatedPattern = regex_txt;

        // catch exceptions, since they may happen from user input in the pattern file
        try
        {
            p.Regex = regex_txt.GetCStr();
        }
        catch (const std::exception& ex)
        {
            return new Error(String::FromFormat(
            "include_utils: failed at translating pattern at line '%s', error: '%s'.", line.GetCStr(), ex.what()
            ));
        }

        patterns.emplace_back(p);
    }
    return HError::None();
}

std::vector<String> match_files(const std::vector<String>& files, const std::vector<Pattern> &patterns)
{
    std::vector<String> matches;
    // if a file entry matches no pattern of the include type it should not be included in the list
    // if a file entry matches a pattern of the include type it should be included unless a pattern of exclude type will match it after
    // if a pattern of exclude type exists but a later pattern causes it to be included, it should be included.

    for (const auto& file : files) {
        bool include = false;
        String normalized_file_path = file.LowerUTF8(); // for case insensitivity
        normalized_file_path = normalize_separators_in_string(normalized_file_path);

        for (const auto& pattern : patterns) {
            if (std::regex_search(normalized_file_path.GetCStr(), pattern.Regex)) {
                if (pattern.Type == eInclude) {
                    include = true;
                } else if (pattern.Type == eExclude) {
                    include = false;
                }
            }
        }

        if (include) {
            matches.push_back(file);
        }
    }

    return matches;
}

std::vector<String> read_file(const String &filename, bool verbose)
{
    std::vector<String> lines;
    std::unique_ptr<Stream> in (File::OpenFileRead(filename));
    if (!in)
    {
        if(verbose)
        {
            printf("Warning: pattern file '%s' not found.\n", filename.GetCStr());
        }
        return lines;
    }
    printf("Info: found pattern file '%s'.\n", filename.GetCStr());

    TextStreamReader sr(std::move(in));

    if (sr.EOS())
        return lines;

    do {
        String line = sr.ReadLine();
        if(!(line.IsNullOrSpace() || line.IsEmpty()))
            lines.push_back(line);
    } while(!sr.EOS());

    return lines;
}

HError MatchPatternPaths(const std::vector<String> &input_files, std::vector<String> &output_matches,
    const std::vector<String> &patterns_description)
{
    std::vector<Pattern> patterns {};
    HError err = description_to_patterns(patterns_description, patterns);
    const std::vector<String> matches = match_files(input_files, patterns);
    output_matches.insert(output_matches.end(), matches.begin(), matches.end());

    return err;
}

HError IncludeFiles(const std::vector<String> &input_files, std::vector<String> &output_files,
    const String &parent, const String &include_pattern_file, const bool verbose)
{
    const String ignore_filename = Path::ConcatPaths(parent, include_pattern_file);
    std::vector<String> patterns_description = read_file(ignore_filename, verbose);
    std::vector<String> matches{};
    HError err = MatchPatternPaths(input_files, matches, patterns_description);
    output_files.insert(output_files.end(), matches.begin(), matches.end());

    return err;
}

} // namespace DataUtil
} // namespace AGS

File: Editor/AGS.Editor/Utils/IncludeUtils.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Text.RegularExpressions;

namespace AGS.Editor
{
    /// <summary>
    /// IncludeUtils deal with the include/exclude list, containing patterns
    /// in the .gitignore format, which may instruct to either include or
    /// exclude particular file path.
    /// The code is based on the Python's fnmatch, and a gist demonstrating
    /// its port to C++, see:
    /// https://github.com/python/cpython/blob/main/Lib/fnmatch.py
    /// https://gist.github.com/alco/1869512
    /// NOTE: the level of support for .gitignore format is not clear,
    /// maybe not all of the possible edge cases are handled here.
    /// For the reference, .gitignore specs:
    /// https://git-scm.com/docs/gitignore
    /// </summary>
    public static class IncludeUtils
    {
        public enum PatternType
        {
            Include,
            Exclude
        };

        /// <summary>
        /// MatchOption flags tell how to interpret pattern and item contents.
        /// </summary>
        [Flags]
        public enum MatchOption
        {
            None = 0,
            /// <summary>
            /// Will ignore item's letter case
            /// </summary>
            CaseInsensitive = 0x0001
        }

        /// <summary>
        /// Describes a single match pattern, which instructs to either include
        /// or excluding matching items. Contains a prepared Regex object.
        /// </summary>
        public struct Pattern
        {
            public readonly PatternType Type;
            public readonly Regex Regex;
            public readonly string OriginalPattern;
            public readonly string RegexPattern;

            public Pattern(PatternType type, Regex rx, string originalPattern, string regexPattern)
            {
                Type = type;
                Regex = rx;
                OriginalPattern = originalPattern;
                RegexPattern = regexPattern;
            }
        }

        /// <summary>
        /// Creates a list of Patterns by parsing Stream contents as a text.
        /// </summary>
        public static Pattern[] CreatePatternList(Stream stream, MatchOption option = MatchOption.None)
        {
            var lines = ReadPatternLines(stream);
            return ParsePatterns(lines, option).ToArray();
        }

        /// <summary>
        /// Creates a list of Patterns by parsing a string.
        /// </summary>
        public static Pattern[] CreatePatternList(string str, MatchOption option = MatchOption.None)
        {
            var lines = ReadPatternLines(str);
            return ParsePatterns(lines, option).ToArray();
        }

        /// <summary>
        /// Filters a list of items using a list of patterns.
        /// Returns a filtered list.
        /// </summary>
        public static string[] FilterItemList(string[] items, Pattern[] patterns, MatchOption option = MatchOption.None)
        {
            // if a file entry matches no pattern of the include type it should not be included in the list
            // if a file entry matches a pattern of the include type it should be included unless a pattern of exclude type will match it after
            // if a pattern of exclude type exists but a later pattern causes it to be included, it should be included.
            List<string> matches = new List<string>();
            foreach (string item in items)
            {
                bool include = false;
                string matchItem = item;
                if (option.HasFlag(MatchOption.CaseInsensitive))
                    matchItem = matchItem.ToLowerInvariant();
                matchItem = NormalizePathSeparatorsInItem(matchItem);

                foreach (var pattern in patterns)
                {
                    if (pattern.Regex.IsMatch(matchItem))
                    {
                        if (pattern.Type == PatternType.Include)
                        {
                            include = true;
                        }
                        else if (pattern.Type == PatternType.Exclude)
                        {
                            include = false;
                        }
                    }
                }

                if (include)
                {
                    matches.Add(item);
                }
            }
            return matches.ToArray();
        }

        private static List<string> ReadPatternLines(Stream stream)
        {
            using (TextReader textReader = new StreamReader(stream))
                return ReadPatternLines(textReader);
        }

        private static List<string> ReadPatternLines(string str)
        {
            using (TextReader textReader = new StringReader(str))
                return ReadPatternLines(textReader);
        }

        private static List<string> ReadPatternLines(TextReader textReader)
        {
            List<string> lines = new List<string>();
            for (string line = textReader.ReadLine(); line != null; line = textReader.ReadLine())
            {
                lines.Add(line);
            }
            return lines;
        }

        /// <summary>
        /// Parses the given list of lines and creates a list of Patterns.
        /// </summary>
        private static List<Pattern> ParsePatterns(List<string> lines, MatchOption option)
        {
            List<Pattern> patterns = new List<Pattern>();
            foreach (string line in lines)
            {
                string parseLine = line.Trim();

                // line is empty or a comment
                if (string.IsNullOrEmpty(parseLine) || parseLine.StartsWith("#"))
                    continue;

                PatternType type = PatternType.Include;
                if (parseLine.StartsWith("!"))
                {
                    type = PatternType.Exclude;
                    parseLine = parseLine.Substring(1);
                }

                if (option.HasFlag(MatchOption.CaseInsensitive))
                    parseLine = parseLine.ToLowerInvariant();
                parseLine = NormalizePathSeparatorsInPattern(parseLine);
                string regexString = PatternToRegexString(parseLine, option);

                Regex rx = null;
                try
                {
                    rx = new Regex(regexString, RegexOptions.None);
                }
                catch (Exception)
                {
                }

                if (rx != null)
                {
                    Pattern p = new Pattern(type, rx, parseLine, regexString);
                    patterns.Add(p);
                }
            }
            return patterns;
        }

        /// <summary>
        /// Normalizes path separators in a input filepath item.
        /// </summary>
        private static string NormalizePathSeparatorsInItem(string pattern)
        {
            return pattern.Replace("\\", "/"); // convert to UNIX paths
        }

        /// <summary>
        /// Normalizes path separators in the pattern string.
        /// </summary>
        private static string NormalizePathSeparatorsInPattern(string pattern)
        {
            return pattern.Replace("\\\\", "/"); // convert to UNIX paths
        }

        /// <summary>
        /// Converts a include/exclude pattern string into the regex string,
        /// which may be used to construct a Regex object.
        /// </summary>
        private static string PatternToRegexString(string pattern, MatchOption option)
        {
            if (string.IsNullOrEmpty(pattern))
                return string.Empty;

            StringBuilder result = new StringBuilder();
            // The pattern must match the path section either at the beginning or right after the path separator
            if (!pattern.StartsWith("/"))
            {
                result.Append("(^|/)");
            }

            int i = 0;
            int n = pattern.Length;
            while (i < n)
            {
                char c = pattern[i++];
                if (c == '*')
                {
                    result.Append(".*");
                }
                else if (c == '?')
                {
                    result.Append('.');
                }
                else if (c == '[')
                {
                    int j = i;
                    /*
                     * The following two statements check if the sequence we stumbled
                     * upon is '[]' or '[!]' because those are not valid character
                     * classes.
                     */
                    if (j < n && pattern[j] == '!')
                        ++j;
                    if (j < n && pattern[j] == ']')
                        ++j;
                    /*
                     * Look for the closing ']' right off the bat. If one is not found,
                     * escape the opening '[' and continue.  If it is found, process
                     * the contents of '[...]'.
                     */
                    while (j < n && pattern[j] != ']')
                        ++j;

                    if (j >= n)
                    {
                        result.Append("\\[");
                    }
                    else
                    {
                        String part = pattern.Substring(i, j - i);
                        part.Replace("\\", " \\\\");
                        char first_char = pattern[i];
                        i = j + 1;
                        result.Append("[");
                        if (first_char == '!')
                        {
                            result.Append("^");
                            result.Append(part.Substring(1));
                        }
                        else if (first_char == '^')
                        {
                            result.Append("\\");
                            result.Append(part);
                        }
                        else
                        {
                            result.Append(part);
                        }
                        result.Append("]");
                    }
                }
                else
                {
                    if (char.IsLetterOrDigit(c) || c == '_')
                    {
                        result.Append(c);
                    }
                    else
                    {
                        // The python approach is to escape all characters that are not alphanumeric
                        result.Append("\\");
                        result.Append(c);
                    }
                }
            }

            // The pattern must match the path section either at the end or right before the path separator
            if (!pattern.EndsWith("/"))
            {
                result.Append("($|/)");
            }
            return result.ToString();
        }
    }
}
